kubeadm: Reorder the token packages more logically

luxas opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://github.com/kubernetes/kubernetes/blob/master/CONTRIBUTING.md and developer guide https://github.com/kubernetes/kubernetes/blob/master/docs/devel/development.md
2. If you want *faster* PR reviews, read how: https://github.com/kubernetes/kubernetes/blob/master/docs/devel/faster_reviews.md
3. Follow the instructions for writing a release note: https://github.com/kubernetes/kubernetes/blob/master/docs/devel/pull-requests.md#release-notes
-->

**What this PR does / why we need it**:

In order to be able to implement https://github.com/kubernetes/kubernetes/pull/41417, the token functionality (which now is spread across the codebase), should be in two places: a generic token functions library, which in the future _may_ [move into client-go](https://github.com/kubernetes/kubernetes/pull/41281#discussion_r101357106) in some form, and a package for the token handling against the api server.

This commit has no large functional changes.

```
kubeadm: Aggregate the token functionality in sane packages.
    
 - Factor out token constants to kubeadmconstants.
 - Move cmd/kubeadm/app/util/{,token/}tokens.go
 - Use the token-id, token-secret, etc constants provided by the bootstrapapi package
 - Move cmd/kubeadm/app/master/tokens.go to cmd/kubeadm/app/phases/token/csv.go
    
This refactor basically makes it possible to hook up kubeadm to the BootstrapSigner controller later on
```

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access) 
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`. 
-->
```release-note
NONE
```
@mikedanese @pires @errordeveloper @dmmcquay @jbeda @GheRivero

luxas commented on cmd/kubeadm/app/constants/constants.go:
```diff
@@ -56,4 +56,18 @@ const (
 	MinimumAddressesInServiceSubnet = 10
+
+	// DefaultTokenDuration specifies the default amount of time that a bootstrap token will be valid
+	DefaultTokenDuration = time.Duration(8) * time.Hour
```
This will become `0` (token doesn't expire)

jbeda replied:
Not sure what this comment means. Are you going to change this to 0?  I think that is probably a compatible default.

luxas replied:
Yes, I'm gonna change it to `0` later, just wanna stay away from functional changes here

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/41509)
<!-- Reviewable:end -->

luxas commented on cmd/kubeadm/app/constants/constants.go:
```diff
@@ -56,4 +56,18 @@ const (
+	// CSVTokenBootstrapUser is currently the user the bootstrap token in the .csv file
+	// TODO: This should change to something more official and supported
+	// TODO: Prefix with kubeadm prefix
+	CSVTokenBootstrapUser = "kubeadm-node-csr"
```
This will become a prefix instead: `system:bootstrapper:`+`<tokenid>`

jbeda replied:
is it `system:bootstrapper` or `system:bootstrap`?

luxas replied:
Oh, recalled incorrectly, yes `system:bootstrap:` as we decided in https://github.com/kubernetes/kubernetes/pull/41281

luxas commented on cmd/kubeadm/app/constants/constants.go:
```diff
@@ -56,4 +56,18 @@ const (
+	// TODO: Prefix with kubeadm prefix
+	CSVTokenBootstrapUser = "kubeadm-node-csr"
+	// CSVTokenBootstrapGroup specifies the group the tokens in the .csv file will belong to
+	CSVTokenBootstrapGroup = "kubeadm:kubelet-bootstrap"
```
This will become `system:bootstrappers`

luxas commented:
cc @dgoodwin as well

jbeda commented on cmd/kubeadm/app/cmd/token.go:
```diff
@@ -69,7 +72,7 @@ func NewCmdToken(out io.Writer, errW io.Writer) *cobra.Command {
 	}
 	createCmd.PersistentFlags().DurationVar(&tokenDuration,
-		"ttl", kubeadmutil.DefaultTokenDuration, "The duration before the token is automatically deleted.")
+		"ttl", kubeadmconstants.DefaultTokenDuration, "The duration before the token is automatically deleted.")
```
Can we add to the help text that a ttl of 0 means no expiration.

luxas replied:
Sure, I'll do that

jbeda commented on cmd/kubeadm/app/phases/token/bootstrap.go:
```diff
@@ -0,0 +1,99 @@
+	tokenutil "k8s.io/kubernetes/cmd/kubeadm/app/util/token"
+	"k8s.io/kubernetes/pkg/api/v1"
+	bootstrapapi "k8s.io/kubernetes/pkg/bootstrap/api"
+	"k8s.io/kubernetes/pkg/client/clientset_generated/clientset"
```
I thought we had completed the switchover to client-go?

dmmcquay replied:
Just about to be merge: https://github.com/kubernetes/kubernetes/pull/41104

jbeda commented on cmd/kubeadm/app/phases/token/bootstrap.go:
```diff
@@ -0,0 +1,99 @@
+	data := map[string][]byte{
+		bootstrapapi.BootstrapTokenIDKey:           []byte(d.ID),
+		bootstrapapi.BootstrapTokenSecretKey:       []byte(d.Secret),
+		bootstrapapi.BootstrapTokenUsageSigningKey: []byte("true"),
```
We probably want to create/add something about `BootstrapTokenUsageAuthentication` and set that to true.

luxas replied:
I'll do that in the next PR :+1:

jbeda reviewed (commented):
Looks pretty good.  I'd add the usage for authn now to save having to do another PR, but that can wait.

dmmcquay reviewed (approved):
lgtm

luxas commented:
@jbeda @dmmcquay Rebased after the client-go merge, this should be good to go :+1:

dmmcquay commented:
please squash and I'll lgtm.

luxas commented:
@dmmcquay I'll do it, but I just made two commits so it would be easier for you to look at :smile:

luxas committed:
kubeadm: Aggregate the token functionality in sane packages.
 - Factor out token constants to kubeadmconstants.
 - Move cmd/kubeadm/app/util/{,token/}tokens.go
 - Use the token-id, token-secret, etc constants provided by the bootstrapapi package
 - Move cmd/kubeadm/app/master/tokens.go to cmd/kubeadm/app/phases/token/csv.go

This refactor basically makes it possible to hook up kubeadm to the BootstrapSigner controller later on

dmmcquay commented:
/lgtm
